runsc/fsgofer/fsgofer.go

```diff
@@ -124,30 +127,47 @@ func (a *attachPoint) Attach() (p9.File, error) {
 	if err != nil {
 		return nil, fmt.Errorf("stat file %q, err: %v", a.prefix, err)
 	}
-	mode := syscall.O_RDWR
-	if a.conf.ROMount || (stat.Mode&syscall.S_IFMT) == syscall.S_IFDIR {
-		mode = syscall.O_RDONLY
-	}
 
-	// Open the root directory.
-	f, err := fd.Open(a.prefix, openFlags|mode, 0)
-	if err != nil {
-		return nil, fmt.Errorf("unable to open file %q, err: %v", a.prefix, err)
+	// Hold the file descriptor we are converting into a p9.File
+	var f *fd.FD
+
+	// Apply the S_IFMT bitmask so we can detect file type appropriately
+	switch fmtStat := stat.Mode & syscall.S_IFMT; {
+	case fmtStat == syscall.S_IFSOCK:
+		// Attempt to open a connection. Bubble up the failures.
+		f, err = fd.OpenUnix(a.prefix)
+		if err != nil {
+			return nil, err
+		}
+
+	default:
+		// Default to Read/Write permissions.
+		mode := syscall.O_RDWR
+
+		// If the configuration is Read Only & the mount point is a directory,
+		// set the mode to Read Only.
+		if a.conf.ROMount || fmtStat == syscall.S_IFDIR {
+			mode = syscall.O_RDONLY
+		}
+
+		// Open the mount point & capture the FD.
+		f, err = fd.Open(a.prefix, openFlags|mode, 0)
+		if err != nil {
+			return nil, fmt.Errorf("unable to open file %q, err: %v", a.prefix, err)
+		}
 	}
 
+	// Close the connection if the UDS is already attached.
 	a.attachedMu.Lock()
```

Comment:
Since the code is more complex now, it would be better to move this check to the top so it doesn't need to close the file on error. There is no harm is holding `a.attachedMu` longer.